server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Rights.scala

```diff
@@ -0,0 +1,147 @@
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.collection.mutable
+import scala.jdk.CollectionConverters._
+
+object Right {
+  val UNSUPPORTED: Option[Boolean] = None
+
+  val Administer = Right(MailboxACL.Right.Administer)
+  val Expunge = Right(MailboxACL.Right.PerformExpunge)
+  val Insert = Right(MailboxACL.Right.Insert)
+  val Lookup = Right(MailboxACL.Right.Lookup)
+  val Read = Right(MailboxACL.Right.Read)
+  val Seen = Right(MailboxACL.Right.WriteSeenFlag)
+  val DeleteMessages = Right(MailboxACL.Right.DeleteMessages)
+  val Write = Right(MailboxACL.Right.Write)
+
+  private val allRights = Seq(Administer, Expunge, Insert, Lookup, Read, Seen, DeleteMessages, Write)
+
+  def forRight(right: MailboxACL.Right): Option[Right] = allRights.find(_.right.equals(right))
+
+  def forChar(c: Char): Option[Right] = allRights.find(_.asCharacter == c)
+}
+
+sealed case class Right(right: MailboxACL.Right) {
+  def asCharacter: Char = right.asCharacter
+
+  def toMailboxRight: MailboxACL.Right = right
+}
+
+object Rights {
+  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[Rights])
+
+  val EMPTY = new Rights(Map())
+
+  class Builder() {
+    private var rights: Seq[(Username, Right)] = Seq()
+
+    def delegateTo(username: Username, rights: Right): Rights.Builder = {
+      this.rights = this.rights :+ (username, rights)
+      this
+    }
+
+    def delegateTo(username: Username, rights: Seq[Right]): Rights.Builder = {
+      this.rights = this.rights ++ rights.map((username, _))
+      this
+    }
+
+    def combine(builder: Rights.Builder): Rights.Builder = {
+      this.rights = this.rights ++ builder.rights
+      this
+    }
+
+    def build: Rights = new Rights(rights.toSeq
+      .groupMap(_._1)(_._2))
+  }
+
+  def builder: Rights.Builder = new Rights.Builder
+
+  def fromACL(acl: MailboxACL): Rights = acl.getEntries.asScala
+    .filter {
+      case (entryKey, _) => isSupported(entryKey)
+    }
+    .map(toRightsBuilder _)
+    .iterator
+    .reduceOption(_ combine _)
+    .getOrElse(builder)
+    .build
+
+  private def toRightsBuilder(entry: (MailboxACL.EntryKey, MailboxACL.Rfc4314Rights)): Rights.Builder = builder
+    .delegateTo(Username.of(entry._1.getName), fromACL(entry._2))
+
+  private def fromACL(rights: MailboxACL.Rfc4314Rights): Seq[Right] = rights.list.asScala
+      .toSeq
+      .flatMap(Right.forRight)
+
+  private def isSupported(key: MailboxACL.EntryKey): Boolean = {
+    if (key.isNegative) {
+      LOGGER.info("Negative keys are not supported")
+      return false
+    }
+    if (key == MailboxACL.OWNER_KEY) {
+      return false
+    }
+    if (key.getNameType ne MailboxACL.NameType.user) {
+      LOGGER.info("{} is not supported. Only 'user' is.", key.getNameType)
+      return false
+    }
+    true
+  }
+}
+
+case class Rights private(rights: Map[Username, Seq[Right]]) {
+
+  def removeEntriesFor(username: Username) = new Rights(rights.filter(!_._1.equals(username)))
```

Comment:
Not an efficient way to remove an element from a map. you can use `rights - username` instead of filter